test/e2e/runner/benchmark.go

```diff
@@ -0,0 +1,185 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	e2e "github.com/tendermint/tendermint/test/e2e/pkg"
+	"github.com/tendermint/tendermint/types"
+)
+
+// Benchmark is a simple function for fetching, calculating and printing
+// the following metrics:
+// 1. Average block production time
+// 2. Block interval standard deviation
+// 3. Max block interval (slowest block)
+// 4. Min block interval (fastest block)
+func Benchmark(testnet *e2e.Testnet, benchmarkLength int64) error {
+
+	block, _, err := waitForHeight(testnet, 0)
+	if err != nil {
+		return err
+	}
+
+	logger.Info("Beginning benchmark period ...", "height", block.Height)
+
+	// wait for the length of the benchmark period in blocks to pass. We allow 5 seconds for each block
+	// which should be sufficient.
+	waitingTime := time.Duration(benchmarkLength*5) * time.Second
+	endHeight, err := waitForAllNodes(testnet, block.Height+benchmarkLength, waitingTime)
+	if err != nil {
+		return err
+	}
+
+	logger.Info("Ending benchmark period.", "height", endHeight)
+
+	// fetch a sample of blocks
+	blocks, err := fetchBlockChainSample(testnet, benchmarkLength)
+	if err != nil {
+		return err
+	}
+
+	// slice into time intervals and collate data
+	timeIntervals := splitIntoBlockIntervals(blocks)
+	testnetStats := extractTestnetStats(timeIntervals)
+	testnetStats.from = blocks[0].Header.Height
+	testnetStats.to = blocks[len(blocks)-1].Header.Height
+
+	// print and return
+	logger.Info(testnetStats.String())
+	return nil
+}
+
+type testnetStats struct {
+	from int64
+	to   int64
+	mean time.Duration
+	std  float64
+	max  time.Duration
+	min  time.Duration
+}
+
+func (t *testnetStats) String() string {
+	return fmt.Sprintf(`Benchmarked from height %v to %v
+	Mean Block Interval: %v
+	Standard Deviation: %f
+	Max Block Interval: %v
+	Min Block Interval: %v
+	`,
+		t.from,
+		t.to,
+		t.mean,
+		t.std,
+		t.max,
+		t.min,
+	)
+}
+
+// fetchBlockChain fetches a complete, up-to-date block history from
+// the freshest testnet archive node.
```

Comment:
what's "freshest"?

Reply:
Sorry this was just directly copied from another part of the e2e tests. Will reword it. It basically means fetch from the first archive node